lib/Sema/TypeCheckConstraints.cpp

```diff
@@ -1131,7 +1138,49 @@ TypeExpr *PreCheckExpression::simplifyTypeExpr(Expr *E) {
                                         ResultTypeRepr);
     return new (TC.Context) TypeExpr(TypeLoc(NewTypeRepr, Type()));
   }
-
+  
+  // Fold 'P & Q' into a composition type
+  if (auto *binaryExpr = dyn_cast<BinaryExpr>(E))
+    if (auto d = dyn_cast<OverloadedDeclRefExpr>(binaryExpr->getFn()))
+      for (auto decl : d->getDecls()) {
```

Comment:
It seems problematic that we have an OverloadedDeclRefExpr here. We don't actually want to be doing overload resolution at all, and what if it comes back with no results for whatever reason?